### DISCUSSION TITLE
Shape.node is assumed as SVGGElement.

### DISCUSSION THREAD

**Discussion by reuse-ay** [maxGraph/maxGraph#536] [Q&A]
I am rewriting some prototype-based code that was written using mxGraph into class-based code using maxGraph. The original code is based on graph editor. There was a problem while rewriting.

In [Graph Editor](https://jgraph.github.io/mxgraph/javascript/examples/grapheditor/www/index.html), there is a `div` element with class `geBackgroundPage`. This is the area where the grid is displayed. In mxgraph, operation is executed in [`validateBackGroundPage`](https://github.com/jgraph/mxgraph/blob/ff141aab158417bd866e2dfebd06c61d40773cd2/javascript/examples/grapheditor/www/js/Editor.js#L2304). Furthermore, `geBackgroundPage` is styled within [`validateBackGroundStyles`](https://github.com/jgraph/mxgraph/blob/ff141aab158417bd866e2dfebd06c61d40773cd2/javascript/examples/grapheditor/www/js/Editor.js#L2395). 

Similar code doesn't work with maxGraph, i.e. the styles aren't applied. In `validateBackGroundStyles`, styles are applies like `graph.view.backgroundPageShape.node.style.backgroundPosition = position;`. But this doesn't make sense, since you can't apply styles to `<g>` elements. In maxGraph, the type of Shape.node is SVGGElement, and some methods have been changed based on that. Related to this issue are `constructor`, `init`, `create` in Shape.

What can I do? Is changing signature allowed by PR?

Sorry. To be precise, it is not that you cannot apply styles to g elements, but rather that when you apply styles to a g element, the styles are applied to the g element's children, not to the g element itself. However, geBackgroundPage does not have any child elements, so the styles have no effect.

**Comment by tbouffard** [comment]
Hi, Here are some comparisons I've made between mxGraph and maxGraph. The analysis seems correct 👍🏿 

In typed-mxgraph, mxShape https://github.com/typed-mxgraph/typed-mxgraph/blob/187dd4f0dc7644c0cfbc998dae5fc90879597d81/lib/shape/mxShape.d.ts
- node: HTMLElement 
- init(container: Element): void;
- create(container: Element): Element; in the mxGraph code, the container parameter was not pass in the underlying createXX methods (https://github.com/jgraph/mxgraph/blob/v4.2.2/javascript/src/js/shape/mxShape.js#L304)

In maxGraph
- node: SVGGElement;
- init(container: HTMLElement | SVGElement) no specified return type so it is inferred to void
- create() no parameter, no specified return type so it is inferred to SVGElement

@reuse-ay just to be sure, are you willing to create a PR to change the signature of methods in Shape (this is what I understood)? or also in other places of the code?

**Reply by reuse-ay** [reply]
Yes, I am willing. Probably fix  Shape's logic not only signature.

**Reply by reuse-ay** [reply]
In [mxShape](https://github.com/jgraph/mxgraph/blob/ff141aab158417bd866e2dfebd06c61d40773cd2/javascript/src/js/shape/mxShape.js#L117), initial value of `node` is `null`. Should type of  `node` be `SVGGElement|HtmlElement|null`? 

- If null should be included in `node`, there are more areas to fix. Probably Mainly null check.

- If null should not be included, fixing gets a little complicated. Now `node` in Shape is initialized by `create` in constructor. However after fixing, `create` method need container argument. So `node` cant be initialized by `create` in constructor. What should I do instead?

**Reply by tbouffard** [reply]
For the type, I suggest not to use `SVGGElement|HtmlElement` but the generic `Element` type if possible (I didn't check what the impacts could be) as we don't effectively create `HtmlElement` instance
For the `null` value, I don't have time to have a look at it right now, I will try to give it a chance later this week.
